### ISSUE TITLE
GPU race conditions from `tf.map_fn` 

### ISSUE THREAD

**Issue by ShellRox** [tensorflow/tensorflow#31449]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): All the code that caused this issue uses Tensorflow/Keras operations.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.14.0
- Python version: 3.6.8
- CUDA/cuDNN version: 10.0/7.6.2
- GPU model and memory: RTX 6000x2, 48 GB

**Describe the current behavior**

I've created a custom layer called `ROI` in Keras that uses `tf.map_fn`, [precisely because it has unknown parameter that it needs to take as tensor object](https://github.com/keras-team/keras/issues/12139). 

This layer works perfectly on CPU inference and training, it also works perfectly on GPU during inference. But during training with a powerful GPU, exception about GPU colocation of `ROI` layer occurs:

```
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot assign a device for operation ROI/map/while/Identity_1: Could not satisfy explicit device specification '' because the node node ROI/map/while/Identity_1 (defined at /path/to/custom/layer/custom.py:70) placed on device No device assignments were active during op 'ROI/map/while/Identity_1' creation.
```

```
[[node ROI/map/while/Identity_1 (defined at /path/to/custom/layer/custom.py:70) Additional information about colocations: No node-device colocations were active during                                                                                                                                                              op 'ROI/map/while/Identity_1' creation.
No device assignments were active during op 'ROI/map/while/Identity_1' creation.`
```

Manual colocation of `ROI` layer to CPU device with `tf.device` worked, but I want `ROI` to support GPU as well.

**My hypothesis**

`ROI` layer works on CPU because only single core at a time should handle the layer - even if multiprocessing is activated, there are few cores slowly balancing the task. 

But whenever GPU is utilized, thousands of cores are working together in parallel and they are not waiting for each other to finish their tasks. Thus one of the processes tries to gather data from `TensorArray` that is still in while loop (using `tf.map_fn`), which causes the error.

**Describe the expected behavior**

Tensorflow should be able to handle these race conditions by waiting for its own `tf.map_fn` to be finished instead of raising exceptions. 

**Code to reproduce the issue**

[This is the code
](https://gist.github.com/ShellRox/a1c11565ed53f219bc1033164947aa72) that instantly causes the mentioned issue on my local machine.

**Other info / logs**

Full Log:

```
Traceback (most recent call last):
  File "/path/to/site-packages/tensorflow/python/client/session.py", line 1356, in _do_call
    return fn(*args)
  File "/path/to/site-packages/tensorflow/python/client/session.py", line 1339, in _run_fn
    self._extend_graph()
  File "/path/to/site-packages/tensorflow/python/client/session.py", line 1374, in _extend_graph
    tf_session.ExtendSession(self._session)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot assign a device for operation ROI/map/while/Identity_1: Could not satisfy explicit device specification '' because the node {{colocation_node ROI/map/while/Identity_1}} was colocated with a group of nodes that required incompatible device '/job:localhost/replica:0/task:0/device:GPU:0'. All available devices [/job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:XLA_GPU:0, /job:localhost/replica:0/task:0/device:XLA_CPU:0, /job:localhost/replica:0/task:0/device:GPU:0].
Colocation Debug Info:
Colocation group had the following types and supported devices:
Root Member(assigned_device_name_index_=1 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' supported_device_types_=[CPU] possible_devices_=[]
StridedSliceGrad: GPU CPU XLA_CPU XLA_GPU
NextIteration: GPU CPU XLA_CPU XLA_GPU
Mul: GPU CPU XLA_CPU XLA_GPU
Equal: GPU CPU XLA_CPU XLA_GPU
DynamicStitch: GPU CPU XLA_CPU XLA_GPU
Fill: GPU CPU XLA_CPU XLA_GPU
FloorMod: GPU CPU XLA_CPU XLA_GPU
Shape: GPU CPU XLA_CPU XLA_GPU
Reshape: GPU CPU XLA_CPU XLA_GPU
TensorArrayReadV3: GPU CPU XLA_CPU XLA_GPU
TensorArrayScatterV3: GPU CPU XLA_CPU XLA_GPU
TensorArraySizeV3: GPU CPU XLA_CPU XLA_GPU
Const: GPU CPU XLA_CPU XLA_GPU
TensorArrayWriteV3: GPU CPU XLA_CPU XLA_GPU
Identity: GPU CPU XLA_CPU XLA_GPU
GreaterEqual: GPU CPU XLA_CPU XLA_GPU
Exit: GPU CPU XLA_CPU XLA_GPU
Cast: GPU CPU XLA_CPU XLA_GPU
ControlTrigger: GPU CPU XLA_CPU XLA_GPU
TensorArrayGradV3: GPU CPU XLA_CPU XLA_GPU
Pack: GPU CPU XLA_CPU XLA_GPU
Enter: GPU CPU XLA_CPU XLA_GPU
TensorArrayV3: GPU CPU XLA_CPU XLA_GPU
Merge: GPU CPU XLA_CPU XLA_GPU
StackV2: GPU CPU XLA_CPU XLA_GPU
Range: GPU CPU XLA_CPU XLA_GPU
TensorArrayGatherV3: GPU CPU XLA_CPU XLA_GPU
StackPushV2: GPU CPU XLA_CPU XLA_GPU
Switch: GPU CPU XLA_CPU XLA_GPU
RealDiv: GPU CPU XLA_CPU XLA_GPU
Add: GPU CPU XLA_CPU XLA_GPU
StridedSlice: GPU CPU XLA_CPU XLA_GPU
Max: GPU CPU XLA_CPU XLA_GPU
LoopCond: GPU CPU XLA_CPU XLA_GPU
Sum: GPU CPU XLA_CPU XLA_GPU
StackPopV2: GPU CPU XLA_CPU XLA_GPU
Sub: GPU CPU XLA_CPU XLA_GPU

Colocation members, user-requested devices, and framework assigned devices, if any:
  ROI/map/TensorArray_2 (TensorArrayV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  ROI/map/while/Identity_1 (Identity)
  ROI/map/while/map/TensorArray_1 (TensorArrayV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  ROI/map/while/map/while/Identity_1 (Identity)
  ROI/map/while/map/while/strided_slice_4/stack (Pack)
  ROI/map/while/map/while/strided_slice_4/stack_1 (Pack)
  ROI/map/while/map/while/strided_slice_4 (StridedSlice)
  ROI/map/while/map/while/Max (Max)
  ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3/Enter (Enter)
  ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3 (TensorArrayWriteV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  ROI/map/while/map/while/Exit_2 (Exit)
  ROI/map/while/map/TensorArrayStack/TensorArraySizeV3 (TensorArraySizeV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  ROI/map/while/map/TensorArrayStack/range/start (Const)
  ROI/map/while/map/TensorArrayStack/range/delta (Const)
  ROI/map/while/map/TensorArrayStack/range (Range)
  ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3 (TensorArrayGatherV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  ROI/map/while/TensorArrayWrite/TensorArrayWriteV3/Enter (Enter)
  ROI/map/while/TensorArrayWrite/TensorArrayWriteV3 (TensorArrayWriteV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  ROI/map/TensorArrayStack/TensorArraySizeV3 (TensorArraySizeV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  ROI/map/TensorArrayStack/range/start (Const)
  ROI/map/TensorArrayStack/range/delta (Const)
  ROI/map/TensorArrayStack/range (Range)
  ROI/map/TensorArrayStack/TensorArrayGatherV3 (TensorArrayGatherV3)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  training/MultiplierWrapper/gradients/f_count_3 (Const)
  training/MultiplierWrapper/gradients/f_count_4 (Enter)
  training/MultiplierWrapper/gradients/Merge_2 (Merge)
  training/MultiplierWrapper/gradients/Switch_2 (Switch)
  training/MultiplierWrapper/gradients/Add_1/y (Const)
  training/MultiplierWrapper/gradients/Add_1 (Add)
  training/MultiplierWrapper/gradients/f_count_5 (Exit)
  training/MultiplierWrapper/gradients/Const (Const)
  training/MultiplierWrapper/gradients/f_acc (StackV2)
  training/MultiplierWrapper/gradients/Enter (Enter)
  training/MultiplierWrapper/gradients/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/b_count_4 (Const)
  training/MultiplierWrapper/gradients/b_count_5 (Enter)
  training/MultiplierWrapper/gradients/Merge_3 (Merge)
  training/MultiplierWrapper/gradients/GreaterEqual_1/Enter (Enter)
  training/MultiplierWrapper/gradients/GreaterEqual_1 (GreaterEqual)
  training/MultiplierWrapper/gradients/b_count_6 (LoopCond)
  training/MultiplierWrapper/gradients/Switch_3 (Switch)
  training/MultiplierWrapper/gradients/Sub_1 (Sub)
  training/MultiplierWrapper/gradients/b_count_7 (Exit)
  training/MultiplierWrapper/gradients/ROI/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3 (TensorArrayGradV3)
  training/MultiplierWrapper/gradients/ROI/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/gradient_flow (Identity)
  training/MultiplierWrapper/gradients/ROI/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3 (TensorArrayScatterV3)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3 (TensorArrayGradV3)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/gradient_flow (Identity)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3 (TensorArrayReadV3)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/Const_1 (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/f_acc_1 (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPushV2_1 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPopV2_1/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPopV2_1 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3 (TensorArrayGradV3)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/gradient_flow (Identity)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayScatter/TensorArrayScatterV3 (TensorArrayScatterV3)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Exit_2_grad/b_exit (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/b_sync (ControlTrigger)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/TensorArrayGradV3 (TensorArrayGradV3)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayGrad/gradient_flow (Identity)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/StackPopV2/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3 (TensorArrayReadV3)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Shape (Shape)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Size (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/add/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/add (Add)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/mod (FloorMod)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Shape_1 (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/range/start (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/range/delta (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/range (Range)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Fill/value (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Fill (Fill)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/StackPopV2/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/DynamicStitch (DynamicStitch)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/StackPopV2/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape (Reshape)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape_1 (Reshape)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/StackPopV2/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Equal (Equal)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Cast (Cast)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Sum (Sum)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/Reshape_2 (Reshape)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/truediv (RealDiv)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/Max_grad/mul (Mul)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/Shape (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/Const (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/f_acc (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPushV2 (StackPushV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPopV2/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPopV2/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPopV2 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/Const_1 (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/f_acc_1 (StackV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/Enter_2 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/Enter_3 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPushV2_1 (StackPushV2)
  training/MultiplierWrapper/gradients/NextIteration_2 (NextIteration)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPopV2_1/Enter (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPopV2_1/Enter_1 (Enter)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/StackPopV2_1 (StackPopV2)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/b_sync (ControlTrigger)
  training/MultiplierWrapper/gradients/NextIteration_3 (NextIteration)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad/Const_2 (Const)
  training/MultiplierWrapper/gradients/ROI/map/while/map/while/strided_slice_4_grad/StridedSliceGrad (StridedSliceGrad)

         [[{{node ROI/map/while/Identity_1}}]]
```

**Comment by oanush** [comment]
Could not reproduce the issue with normal CPU and GPU run-time, also not able to reproduce issue for the GPU model mentioned.

**Comment by eherrerosj** [comment]
We are working on reproducing it through a Google Colab notebook to share with you. Not sure if the issue appears only with powerful multi GPUs or also with average GPUs like the default K80 from colab. We will share updates ASAP @oanush, @jvishnuvardhan

**Comment by ShellRox** [comment]
# Update

Tried to implement the slightly altered code to Google Colab and couldn't replicate the issue. Then tried the same slightly altered code on my machine and surprisingly the issue still couldn't be replicated, **before** I pre-loaded the weights before training - then issue immediately occurred back up (on my local machine, and not colab).

Thus there are two prerequisites to replicate this issue:

- Load **arbitrary** of `.h5` weights before compiling the network with `ROI` layer implemented within.
- Execute the training script on a powerful GPU. K80 from Colab fails to replicate this issue for some reason, but I've tested it on RTX 6000x2, P5000, P4000 and GTX 960 and they successfully replicate the issue if condition in the first bullet point is satisfied (although GTX 960 might fail sometimes when it's overloaded).

By the way, I highlighted word **arbitrary** in the first bullet point because this happens for any weight. Thus you can simply train the network, save its weights and then try training again by pre-loading weights with `model.load_weights` before `model.compile`.

---

Currently I'm minimizing the code as much as possible, and I believe I'll be able to send it tomorrow. I think it'll make it easier to replicate this issue (although it is important to consider the usage of powerful GPU). I'll be probably able to post minimized code by tomorrow.

Otherwise, I'm confident that satisfying mentioned prerequisites will replicate this issue.

Thank you for the response! @oanush @jvishnuvardhan

**Comment by ShellRox** [comment]
> Could not reproduce the issue with normal CPU and GPU run-time, also not able to reproduce issue for the GPU model mentioned.

I've been able to construct a minimal code that causes this issue instantly on my local machine, [you can see it here](https://gist.github.com/ShellRox/a1c11565ed53f219bc1033164947aa72) (I've also updated the issue with this link). Please make sure to execute this script **twice**, i.e `python GPU_ROI_issue.py;python GPU_ROI_issue.py` so that it saves the weights after training and then tries to train with loaded weights which causes the error mentioned in this issue. 

Also as mentioned earlier, please make sure that this code is executed on efficient GPU, since from my experience it is only caused on them.

Thank you!

**Comment by ShellRox** [comment]
# Update

No fix is yet found for training a model containing layer that uses `tf.map_fn` on GPU with pre-loaded weights. As soon as any weights of any layers are manually set (i.e with `layer.set_weights` or `model.load_weights`), which uses `batch_set_value` to pass them into the `feed_dict` to a symbolic callable function that Keras creates (but kernel/bias initializers of layers work perfectly though).

Perhaps `channels_first` (NCHW) setting of layers might have something to do with it, considering that Tensorflow has a default channel order of NHWC and therefore is probably optimized for that exact format.

The only workaround with `tf.map_fn` is, by colocating `ROIPoolingLayer` on CPU device with `tf.device("CPU:0")`, but this is obviously not an optimal solution (considering that GPU support is highly significant for efficiency).

Another obvious workaround is using vectorized operations only, but this might not be an easy task for such layer without hardcoding.

**Comment by eherrerosj** [comment]
@oanush @jvishnuvardhan were you able to reproduce the issue?

> > Could not reproduce the issue with normal CPU and GPU run-time, also not able to reproduce issue for the GPU model mentioned.
> 
> I've been able to construct a minimal code that causes this issue instantly on my local machine, [you can see it here](https://gist.github.com/ShellRox/a1c11565ed53f219bc1033164947aa72) (I've also updated the issue with this link). Please make sure to execute this script **twice**, i.e `python GPU_ROI_issue.py;python GPU_ROI_issue.py` so that it saves the weights after training and then tries to train with loaded weights which causes the error mentioned in this issue.
> 
> Also as mentioned earlier, please make sure that this code is executed on efficient GPU, since from my experience it is only caused on them.
> 
> Thank you!

**Comment by ShellRox** [comment]
> # Update
> No fix is yet found for training a model containing layer that uses `tf.map_fn` on GPU with pre-loaded weights. As soon as any weights of any layers are manually set (i.e with `layer.set_weights` or `model.load_weights`), which uses `batch_set_value` to pass them into the `feed_dict` to a symbolic callable function that Keras creates (but kernel/bias initializers of layers work perfectly though).
> 
> Perhaps `channels_first` (NCHW) setting of layers might have something to do with it, considering that Tensorflow has a default channel order of NHWC and therefore is probably optimized for that exact format.
> 
> The only workaround with `tf.map_fn` is, by colocating `ROIPoolingLayer` on CPU device with `tf.device("CPU:0")`, but this is obviously not an optimal solution (considering that GPU support is highly significant for efficiency).
> 
> Another obvious workaround is using vectorized operations only, but this might not be an easy task for such layer without hardcoding.

@oanush @jvishnuvardhan Channel order is not a problem, I've set all layers to `channels_last` (NHWC) order and modified ROI pooling layer to support this change. I've also replaced every `keras` usage with `tf.keras` (to ensure this is not a compatibility issue) - and I can still reproduce this error.

[This is the code using tf.keras and NHWC layer order that reproduces the issue mentioned in the original post](https://gist.github.com/ShellRox/91bead44f6b476a08b9a3c223629ce4c)

**Comment by ekuznetsov139** [comment]
I can reproduce it with a 1080 Ti.

The problem is not a race condition. Part of the constructed graph is not GPU-friendly and it has to be placed on the CPU. The placer manages to work out a good arrangement if the model has not been loaded any weights. Otherwise, something somewhere takes the wrong turn and the placer arrives to a pathological state. 

I tried to compare graph dumps with and without weight load, but they are exceedingly complicated and the reason for the mismatch is not apparent. Although it looks like the run with weight load is missing all the backprop-related nodes until a late point. For example, without weight load, placer_input_2.pbtxt is 325 kb and it has lots of "training/..." nodes. With weight load, placer_input_2.pbtxt and it has no "training" nodes at all. (They only appear in placer_input_4.pbtxt.) In placer_input_2.pbtxt, the node roi_layer/map/while/TensorArrayWrite/TensorArrayWriteV3 gets assigned to the CPU without weight load and to the GPU with weight load.

**Comment by ShellRox** [comment]
> I can reproduce it with a 1080 Ti.
> 
> The problem is not a race condition. Part of the constructed graph is not GPU-friendly and it has to be placed on the CPU. The placer manages to work out a good arrangement if the model has not been loaded any weights. Otherwise, something somewhere takes the wrong turn and the placer arrives to a pathological state.
> 
> I tried to compare graph dumps with and without weight load, but they are exceedingly complicated and the reason for the mismatch is not apparent. Although it looks like the run with weight load is missing all the backprop-related nodes until a late point. For example, without weight load, placer_input_2.pbtxt is 325 kb and it has lots of "training/..." nodes. With weight load, placer_input_2.pbtxt is 116 kb and it has no "training" nodes at all. (They only appear in placer_input_4.pbtxt.) In placer_output_2.pbtxt, the node roi_layer/map/while/TensorArrayWrite/TensorArrayWriteV3 gets assigned to the CPU without weight load and to the GPU with weight load.

Hello, thank you for the attempt of reproduction and debugging info. I think race conditions or some GPU related issues might be precisely caused due to part of constructed graph not being GPU-friendly as you mentioned (I assume this implies `tf.map_fn` compatibility, or at least some part of its code), but I'm not aware of exact issue.

Manual colocation (`tf.device` wrapper) completely fixes this issue by colocating `ROI` graph to CPU - although I don't think this is as efficient as Tensorflow's automatic colocator. But I still find weight loading concept quite perplexing (and its relations with device colocations), thank you very much for the experiment though, it still has very interesting results.

**Comment by ekuznetsov139** [comment]
More information:

* During placement, in core/common_runtime/colocation_graph.cc, placer creates node 93
```
{name:'roi_layer/map/while/map/TensorArray_1' id:93 op device:{/job:localhost/replica:0/task:0/device:GPU:0} def:{{{node roi_layer/map/while/map/TensorArray_1}} = TensorArrayV3[clear_after_read=
true, dtype=DT_FLOAT, dynamic_size=false, element_shape=<unknown>, identical_element_shapes=true, tensor_array_name=""](roi_layer/map/while/map/strided_slice)}}
```

and assigns it to a Member 

```
(Member(assigned_device
_name_index_=-1 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='' supported_device_types_=[GPU,  CPU, XLA_CPU, XLA_GPU] possible_devices_=[])

```

And it assigns node 588
```
{name:'training/SGD/gradients/roi_layer/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPushV2' id:588 op device:{} def:{{{node training/SGD/gradie
nts/roi_layer/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPushV2}} = StackPushV2[T=DT_RESOURCE, _class=["loc:@roi_layer/map/while/map/TensorArray_1"], swap_me
mory=false](training/SGD/gradients/roi_layer/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/Enter, roi_layer/map/while/map/TensorArray_1, ^training/SGD/gradients/Add)
}}
```

to member 

```
(Member(assigned_device_name_index_=-1 requested_device_name_='' assigned_device_name_='' resource_device_name_='' supported_device_types_=[CPU] possible_devices_=[])```

- so far so good.

Then the placer does a lot of optimizations and eventually gets around to colocate node 93 and node 588:

```
Colocating 93 {name:'roi_layer/map/while/map/TensorArray_1' id:93 op device:{/job:localhost/replica:0/task:0/device:GPU:0} def:{{{node roi_layer/map/while/map/TensorArray_1}} = TensorArrayV3[clear_after_read=true, dtype=DT_FLOAT, dynamic_size=false, element_shape=<unknown>, identical_element_shapes=true, tensor_array_name=""](roi_layer/map/while/map/strided_slice)}} and 588 {name:'training/SGD/gradients/roi_layer/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPushV2' id:588 op device:{} def:{{{node training/SGD/gradients/roi_layer/map/while/map/TensorArrayStack/TensorArray
GatherV3_grad/TensorArrayGrad/TensorArrayGradV3/StackPushV2}} = StackPushV2[T=DT_RESOURCE, _class=["loc:@roi_layer/map/while/map/TensorArray_1"], swap_memory=false](training/SGD/gradients/roi_layer/map/while/map/TensorArrayStack/TensorArrayGatherV3_grad/TensorArrayGrad/TensorArrayGradV3/Enter, roi_layer/map/while/map/TensorArray_1, ^training/SGD/gradients/Add)}}
New root: 0x55a18977e448 Member(assigned_device_name_index_=-1 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resourc
e_device_name_='' supported_device_types_=[GPU, CPU, XLA_CPU, XLA_GPU] possible_devices_=[]
MergeDeviceNames(0x55a18977e448 0x55a1897bcdb0): requested /job:localhost/replica:0/task:0/device:GPU:0
0x55a18977e448: merging devices with 0x55a1897bcdb0
New root: 0x55a18977e448 Member(assigned_device_name_index_=-1 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='' supported_device_types_=[CPU] possible_devices_=[]
```

And now we end up with a Member object that was _requested_ to be on the GPU, but can only be _placed_ on the CPU. (Because colocation logic does not check for this possibility.)

Then, quite a while later, the problem is noticed and reported in a singularly unhelpful error message.

To me this looks like a bug in the placer.

**Comment by ekuznetsov139** [comment]
Okay, I think I see the big picture.
We have a graph containing an operation that can be placed on the GPU, but whose gradient can only be placed on the CPU.

When constructed normally via model.compile/model.fit, it works because the graph is completely constructed in one go.

When we load weights, the graph is partially constructed inside K.batch_set_value, using only a subset of the nodes. Since gradient ops are still missing, the placer has no problem placing everything on the GPU.

When we then try to train, we add the remaining ops, but we retain the device designations made during the partial construction, which are incompatible with the full construction.

Or something like that. :)

**Comment by ShellRox** [comment]
> Okay, I think I see the big picture.
> We have a graph containing an operation that can be placed on the GPU, but whose gradient can only be placed on the CPU.
> 
> When constructed normally via model.compile/model.fit, it works because the graph is completely constructed in one go.
> 
> When we load weights, the graph is partially constructed inside K.batch_set_value, using only a subset of the nodes. Since gradient ops are still missing, the placer has no problem placing everything on the GPU.
> 
> When we then try to train, we add the remaining ops, but we retain the device designations made during the partial construction, which are incompatible with the full construction.
> 
> Or something like that. :)
> 
> P.S. And I found where exactly device designations are being retained. Right here:
> https://github.com/tensorflow/tensorflow/blob/ee16fcac960ae660e0e4496658a366e2f745e1f0/tensorflow/core/common_runtime/graph_execution_state.cc#L243
> 
> After commenting out that line as well as these
> 
> https://github.com/tensorflow/tensorflow/blob/ee16fcac960ae660e0e4496658a366e2f745e1f0/tensorflow/core/common_runtime/direct_session.cc#L1502-L1505
> 
> , and then rebuilding, the error goes away!
> ( Of course, those lines are probably there for a reason and simply axing them in the live repository is probably not the optimal solution ... )

That's very interesting, so subset of regular operations that are temporarily held with `K.batch_set_value` during weight assignment have complete support for GPU - therefore it is assumed that the whole graph, including the complement set containing gradient operations should be colocated on GPU device (however, not all gradient operations in this complement set have GPU kernels). 

I still don't get why automatic weight initializers and `load_weights` use similar graph construction and colocation schemes though, wouldn't it be easier? (perhaps not, but not sure why).

Regardless, thank you for your findings. I hope there is at least slightly simple general solution for this weird issue.

**Comment by ekuznetsov139** [comment]
> therefore it is assumed that the whole graph, including the complement set containing gradient operations should be colocated on GPU device 

I don't know if it's _assumed_, or simply that no one yet realized this scenario is possible. 

> I still don't get why automatic weight initializers and load_weights use similar graph construction and colocation schemes though, wouldn't it be easier? (perhaps not, but not sure why).

As far as I understand, when the graph is initially constructed, weight tensors are provided with initializers. They are told "if you are asked to execute but you don't have any contents, go to this RNG and pull some values out of it to fill yourself." But nothing actually happens until you start the session run inside model.fit().

When you load weights, you explicitly assign contents to the tensors, which requires a separate session run.

What I'm not sure about is why you don't get the complete graph right out of model.compile(), since that would solve the problem too.

FYI, on a tangential topic, in case you haven't realized. On your system, there are (probably) two separate keras Python source trees, one in "keras", and one in "tensorflow.keras". They aren't necessarily the same (they are very similar but may have different versions.)  If you pull the code from the tensorflow repository right now and follow the instructions to build it from source, you'll get the new binaries, but your code will still be using the same python code.

**Comment by ShellRox** [comment]
> > I still don't get why automatic weight initializers and load_weights use similar graph construction and colocation schemes though, wouldn't it be easier? (perhaps not, but not sure why).
> 
> As far as I understand, when the graph is initially constructed, weight tensors are provided with initializers. They are told "if you are asked to execute but you don't have any contents, go to this RNG and pull some values out of it to fill yourself." But nothing actually happens until you start the session run inside model.fit().
> 
> When you load weights, you explicitly assign contents to the tensors, which requires a separate session run.
> 
> What I'm not sure about is why you don't get the complete graph right out of model.compile(), since that would solve the problem too.

Yes, what I mentioned was that it could be more explicit if random initializers created separate session to fill the weight tensors just like manual methods. But I agree, whole graph construction during `model.compile` would fix this issue as well - I guess there's some drawback to that though.



> FYI, on a tangential topic, in case you haven't realized. On your system, there are (probably) two separate keras Python source trees, one in "keras", and one in "tensorflow.keras". They aren't necessarily the same (they are very similar but may have different versions.) If you pull the code from the tensorflow repository right now and follow the instructions to build it from source, you'll get the new binaries, but your code will still be using the same python code.
> 
> To use the "correct" keras, you need to add "tensorflow." to all import directives.
> 
> ```
> import tensorflow.keras as keras
> from tensorflow.keras import callbacks
> import tensorflow.keras.layers as layers
> from tensorflow.keras.activations import linear
> ```
> etc

Yes, I am aware of that. In fact I've linked [a new code](https://gist.github.com/ShellRox/91bead44f6b476a08b9a3c223629ce4c) in one of my replies for this issue, that uses `tf.keras` and NHWC format (I was hoping it would fix this problem but it didn't)


> FYI2, for a hacky workaround for this problem that does not involve manually holding the ROI graph to the CPU, you can run the training for 1 epoch, _then_ load weights, then train for however long you need.

I don't think I understand how can these steps be executed. Can I train the model for `1` epoch and then load weights without stopping this process? I guess I need to temporarily change things in source code?

**Comment by ekuznetsov139** [comment]
```    history = model.fit_generator(TrainSequence(data_set),
                          epochs=1,
                          use_multiprocessing=False,
                          workers=1)
     if os.path.exists(weights_file):
        model.load_weights("new.h5")
     history = model.fit_generator(TrainSequence(data_set),
                          epochs=10,
                          use_multiprocessing=False,
                          workers=1)
```
Pretty sure it would work

**Comment by aaroey** [comment]
@reedwm could you help to take a look at this?

**Comment by hoangcuong2011** [comment]
@ekuznetsov139: I got a similar problem and I confirmed your trick of fit - load_weights and then fit again works. Thank you very much!

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
